From edfe7087fdf56e5f49f899dfc9e5114f4f4c5949 Mon Sep 17 00:00:00 2001
From: Alexey Murz Korepov <MurzNN@gmail.com>
Date: Wed, 13 Nov 2024 14:24:15 +0400
Subject: [PATCH 1/6] Extend the documentation

---
 core/modules/locale/locale.api.php | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/core/modules/locale/locale.api.php b/core/modules/locale/locale.api.php
index c4aa81c223c5..d92b54cbfb7d 100644
--- a/core/modules/locale/locale.api.php
+++ b/core/modules/locale/locale.api.php
@@ -41,6 +41,11 @@
  * @code
  * 'interface translation server pattern': public://translations/%project-%version.%language.po
  * @endcode
+ * And an additional streamwrapper "module://" is available to set the relative
+ * path to the current module:
+ * @code
+ * 'interface translation server pattern': module://translations/%version.%language.po
+ * @endcode
  *
  * Multiple custom modules or themes sharing the same po file should have
  * matching definitions. Such as modules and sub-modules or multiple modules in
-- 
GitLab


From 0a0fe25bc8ce40d2b7c95d14cf5a074e5cb2ccb8 Mon Sep 17 00:00:00 2001
From: Liam Morland <liam@openplus.ca>
Date: Mon, 16 Jun 2025 12:21:09 -0400
Subject: [PATCH 2/6] Improve addition to the documentation

---
 core/modules/locale/locale.api.php | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/core/modules/locale/locale.api.php b/core/modules/locale/locale.api.php
index d92b54cbfb7d..6d949ff9dc2a 100644
--- a/core/modules/locale/locale.api.php
+++ b/core/modules/locale/locale.api.php
@@ -41,8 +41,7 @@
  * @code
  * 'interface translation server pattern': public://translations/%project-%version.%language.po
  * @endcode
- * And an additional streamwrapper "module://" is available to set the relative
- * path to the current module:
+ * Streamwrapper "module://" addresses paths relative to the current project.
  * @code
  * 'interface translation server pattern': module://translations/%version.%language.po
  * @endcode
-- 
GitLab


From a5c78e5a88e2c378ab132c6d310261ba880a3ad1 Mon Sep 17 00:00:00 2001
From: Liam Morland <liam@openplus.ca>
Date: Mon, 16 Jun 2025 12:21:29 -0400
Subject: [PATCH 3/6] Improve documentation for existing relative paths

---
 core/modules/locale/locale.api.php | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/core/modules/locale/locale.api.php b/core/modules/locale/locale.api.php
index 6d949ff9dc2a..6092a73ffe69 100644
--- a/core/modules/locale/locale.api.php
+++ b/core/modules/locale/locale.api.php
@@ -73,8 +73,8 @@
  *   the same project name will be listed as one the translation status list.
  * - "interface translation server pattern": URL of the .po translation files
  *   used to download the files from. The URL contains tokens which will be
- *   replaced by appropriate values. The file can be locate both at a local
- *   relative path, a local absolute path and a remote server location.
+ *   replaced by appropriate values. This value can be a local path relative to
+ *   the Drupal root, a local absolute path, or a remote server location.
  *
  * The following tokens are available for the server pattern:
  * - "%core": Core version. Value example: "8.x".
-- 
GitLab


From 62407b2291fe2b801c34babce4b836899d6b7a8d Mon Sep 17 00:00:00 2001
From: Sylvain VANEL <sylvain.vanel@kleegroup.com>
Date: Wed, 16 Jul 2025 15:09:35 +0200
Subject: [PATCH 4/6] Check the `extension://` prefix in the translation server
 path and replace it with the actual extension path

---
 core/modules/locale/locale.api.php | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/core/modules/locale/locale.api.php b/core/modules/locale/locale.api.php
index 6092a73ffe69..63a69919e02c 100644
--- a/core/modules/locale/locale.api.php
+++ b/core/modules/locale/locale.api.php
@@ -41,9 +41,9 @@
  * @code
  * 'interface translation server pattern': public://translations/%project-%version.%language.po
  * @endcode
- * Streamwrapper "module://" addresses paths relative to the current project.
+ * Streamwrapper "extension://" addresses paths relative to the current project.
  * @code
- * 'interface translation server pattern': module://translations/%version.%language.po
+ * 'interface translation server pattern': extension://translations/%version.%language.po
  * @endcode
  *
  * Multiple custom modules or themes sharing the same po file should have
-- 
GitLab


From a7b1050d09d4aa0d73c130739ed53703daf6d4ca Mon Sep 17 00:00:00 2001
From: Sylvain VANEL <sylvain.vanel@kleegroup.com>
Date: Sat, 4 Jul 2026 13:37:52 +0200
Subject: [PATCH 5/6] Re-apply module:// and extension:// server pattern
 handling on the refactored code

Co-authored-by: Alexey Murz Korepov <MurzNN@gmail.com>
---
 .../locale/src/LocaleProjectRepository.php      | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/core/modules/locale/src/LocaleProjectRepository.php b/core/modules/locale/src/LocaleProjectRepository.php
index b8416d2076d3..eca25bc3a1c8 100644
--- a/core/modules/locale/src/LocaleProjectRepository.php
+++ b/core/modules/locale/src/LocaleProjectRepository.php
@@ -7,6 +7,7 @@
 use Drupal\Core\Cache\Cache;
 use Drupal\Core\Cache\CacheBackendInterface;
 use Drupal\Core\Config\ConfigFactoryInterface;
+use Drupal\Core\Extension\ExtensionPathResolver;
 use Drupal\Core\Extension\ModuleExtensionList;
 use Drupal\Core\Extension\ModuleHandlerInterface;
 use Drupal\Core\Extension\ThemeExtensionList;
@@ -27,6 +28,7 @@ public function __construct(
     protected ModuleExtensionList $moduleExtensionList,
     protected ThemeExtensionList $themeExtensionList,
     protected ConfigFactoryInterface $configFactory,
+    protected ExtensionPathResolver $extensionPathResolver,
   ) {}
 
   /**
@@ -155,6 +157,21 @@ public function buildProjects(): array {
         }
       }
 
+      // Check for the `extension://` prefix in the translation server pattern and
+      // replace it to the current extension path.
+      if (
+        isset($data['info']['interface translation server pattern'])
+        && str_starts_with($data['info']['interface translation server pattern'], 'extension://')
+      ) {
+        $path = $this->extensionPathResolver->getPath($data['project_type'], $name);
+        $data['info']['interface translation server pattern'] =
+          str_replace(
+            'extension://',
+            $path . '/',
+            $data['info']['interface translation server pattern']
+          );
+      }
+
       $refreshed_projects[$name] = new LocaleTranslatableProject(
         name: $name,
         type: $data['project_type'],
-- 
GitLab


From a4d33e82d755175ca4ec1fc35e8f52978bc44dc8 Mon Sep 17 00:00:00 2001
From: Sylvain VANEL <sylvain.vanel@kleegroup.com>
Date: Sat, 4 Jul 2026 13:57:58 +0200
Subject: [PATCH 6/6] Fix PHPCS and PhpUnit

---
 core/modules/locale/src/LocaleProjectRepository.php   |  4 ++--
 .../tests/src/Unit/LocaleProjectRepositoryTest.php    | 11 +++++++++++
 2 files changed, 13 insertions(+), 2 deletions(-)

diff --git a/core/modules/locale/src/LocaleProjectRepository.php b/core/modules/locale/src/LocaleProjectRepository.php
index eca25bc3a1c8..f6ba0267d3f0 100644
--- a/core/modules/locale/src/LocaleProjectRepository.php
+++ b/core/modules/locale/src/LocaleProjectRepository.php
@@ -157,8 +157,8 @@ public function buildProjects(): array {
         }
       }
 
-      // Check for the `extension://` prefix in the translation server pattern and
-      // replace it to the current extension path.
+      // Check for the `extension://` prefix in the translation server pattern
+      // and replace it to the current extension path.
       if (
         isset($data['info']['interface translation server pattern'])
         && str_starts_with($data['info']['interface translation server pattern'], 'extension://')
diff --git a/core/modules/locale/tests/src/Unit/LocaleProjectRepositoryTest.php b/core/modules/locale/tests/src/Unit/LocaleProjectRepositoryTest.php
index 40271b8e5e77..0b8c86999395 100644
--- a/core/modules/locale/tests/src/Unit/LocaleProjectRepositoryTest.php
+++ b/core/modules/locale/tests/src/Unit/LocaleProjectRepositoryTest.php
@@ -6,6 +6,7 @@
 
 use Drupal\Core\Cache\CacheBackendInterface;
 use Drupal\Core\Config\ConfigFactoryInterface;
+use Drupal\Core\Extension\ExtensionPathResolver;
 use Drupal\Core\Extension\ModuleExtensionList;
 use Drupal\Core\Extension\ModuleHandlerInterface;
 use Drupal\Core\Extension\ThemeExtensionList;
@@ -74,6 +75,13 @@ class LocaleProjectRepositoryTest extends UnitTestCase {
    */
   private ConfigFactoryInterface $config;
 
+  /**
+   * The key value memory factory.
+   *
+   * @var \Drupal\Core\Extension\ExtensionPathResolver
+   */
+  private ExtensionPathResolver $extensionPathResolver;
+
   /**
    * {@inheritdoc}
    */
@@ -85,6 +93,7 @@ protected function setUp(): void {
     $this->moduleExtensionList = $this->createStub(ModuleExtensionList::class);
     $this->themeExtensionList = $this->createStub(ThemeExtensionList::class);
     $this->config = $this->createStub(ConfigFactoryInterface::class);
+    $this->extensionPathResolver = $this->createStub(ExtensionPathResolver::class);
     $this->localeProjectRepository = new LocaleProjectRepository(
       $this->cache,
       $this->keyValueMemoryFactory,
@@ -92,6 +101,7 @@ protected function setUp(): void {
       $this->moduleExtensionList,
       $this->themeExtensionList,
       $this->config,
+      $this->extensionPathResolver,
     );
   }
 
@@ -143,6 +153,7 @@ public function testSorting(): void {
       $this->moduleExtensionList,
       $this->themeExtensionList,
       $this->config,
+      $this->extensionPathResolver,
     );
     $z = new LocaleTranslatableProject('z', 'z', 'z', 'z', 'z');
     $this->localeProjectRepository->set($z);
-- 
GitLab
